### PR TITLE
Add pre- and post-hooks

### PR DESCRIPTION
**Pull request by marchuffnagle**
This PR adds support for pre- and post-hook lambda functions. If configured, these have the ability to modify (and possibly short-circuit) the incoming request, and also modify the response from stac-server before it is sent to the client.

The details of this change are documented in [README.md](https://github.com/stac-utils/stac-server/blob/pre-and-post-hooks/README.md#pre--and-post-hooks) and [tests/README.md](https://github.com/stac-utils/stac-server/blob/pre-and-post-hooks/tests/README.md#aws-tests).

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -381,6 +382,49 @@ Stac-server can also be subscribed to SNS Topics that publish complete STAC Item
 
 Errors that occur during ingest will end up in the dead letter processing queue, where they are processed by the `stac-server-<stage>-failed-ingest` Lambda function. Currently all the failed-ingest Lambda does is log the error, see the CloudWatch log `/aws/lambda/stac-server-<stage>-failed-ingest` for errors.
 
+## Pre- and Post-Hooks
+
+Stac-server supports two hooks into the request process: a pre-hook and a post-hook. These are each lambda functions which, if configured, will be invoked by stac-server. It is assumed that the stac-server lambda has been granted permission to invoke these lambda functions, if configured.
+
+### Pre-Hook
+
+If the stac-server is deployed with the `PRE_HOOK` environment variable set to the name of a lambda function, then that function will be called as the pre-hook.
```
Comment (philvarner) [inline README.md:391]: could you add an example of what the value PRE_HOOK and POST_HOOK are set to look like? It's unclear to me just reading this what the "name" of the lambda is.

Comment (philvarner) [inline README.md:391]: could you add an example of what the value PRE_HOOK and POST_HOOK are set to look like? It's unclear to me just reading this what the "name" of the lambda is.

File: CHANGELOG.md [hunk 2]
```diff
@@ -5,29 +5,19 @@ All notable changes to this project will be documented in this file.
...
-- Added STAC_API_VERSION to display on the API landing page within the api.js file
-under src/lib/api.js in the collectionsToCatalogLinks function
+- Added STAC_API_VERSION to display on the API landing page within the api.js file under src/lib/api.js in the collectionsToCatalogLinks function
+- Add support for pre- and post-hooks
 
 ### Changed
 
 - Changed the rel type to 'server' for the URL to the STAC API webpage inside the Links object
+- Modified sortby and collections parameters in nextlink
+- Used map instead of foreach/push in api.js file
```
Comment (philvarner) [inline CHANGELOG.md:20]: Could you add a note in Changed about not compressing the  response and the 6MB limit?

Comment (philvarner) [inline CHANGELOG.md:20]: Could you add a note in Changed about not compressing the  response and the 6MB limit?

File: README.md [hunk 3]
```diff
@@ -381,6 +382,49 @@ Stac-server can also be subscribed to SNS Topics that publish complete STAC Item
...
+  pre-hook -- APIGatewayProxyEvent --> stac-server
+  post-hook[post-hook lambda]
+  stac-server -- APIGatewayProxyResult --> post-hook
+  post-hook -- APIGatewayProxyResult --> client
+```
+
+### Validation
+
+The outputs of the pre- and post-hooks are validated and, if they don't comply with the defined schemas, an internal server error will be returned to the client. Information about the invalid event, as well as details about the parsing errors, will be logged to CloudWatch.
+
```
Comment (philvarner) [inline README.md:427]: maybe also add something here about the 6MB limit for the event size

**Review by philvarner — DISMISSED** [review]

**Review by philvarner — CHANGES_REQUESTED** [review]
Few documentation requests.

**Comment by marchuffnagle** [comment]
> Under what conditions would the responses be compressed?

I _believe_ that API Gateway is configured to compress responses larger than 1 MB. There is a limit of 6MB on the uncompressed response.

**Review by philvarner — APPROVED** [review]

**Pull request merged**